File: Data Structures & Algorithms/implement-prefix-tree/submission-2.py
class TrieNode:
    def __init__(self):
        self.children = [None] * 26
        self.is_end = False

class PrefixTree:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        cur = self.root
        for ch in word:
            i = ord(ch) - ord("a")
            if not cur.children[i]:
                cur.children[i] = TrieNode()
            cur = cur.children[i]
        cur.is_end = True

    def search(self, word: str) -> bool:
        cur = self.root
        for ch in word:
            i = ord(ch) - ord("a")
            if not cur.children[i]:
                return False
            cur = cur.children[i]
        return cur.is_end

    def startsWith(self, prefix: str) -> bool:
        cur = self.root
        for ch in prefix:
            i = ord(ch) - ord("a")
            if not cur.children[i]:
                return False
            cur = cur.children[i]
        return True

File: Data Structures & Algorithms/design-word-search-data-structure/submission-2.cpp
class WordDictionary {
private:
    struct Node {
        Node* children[26]{};
        bool isEnd = false;
    };

    Node* root;

    bool dfs(const string& word, int index, Node* node) {
        if (index == word.size()) return node->isEnd;

        char c = word[index];
        if (c != '.') {
            int i = c - 'a';
            if (!node->children[i]) {
                return false;
            }

            return dfs(word, index + 1, node->children[i]);
        }

        for (Node* child : node->children) {
            if (child && dfs(word, index + 1, child)) {
                return true;
            }
        }

        return false;
    }

public:
    WordDictionary() {
        root = new Node();
    }
    
    void addWord(string word) {
        Node* cur = root;

        for (char c : word) {
            int idx = c - 'a';
            if (!cur->children[idx]) {
                cur->children[idx] = new Node();
            }
            cur = cur->children[idx];
        }

        cur->isEnd = true;
    }
    
    bool search(string word) {
        return dfs(word, 0, root);
    }
};
